File: matlab/examples/gettingstarted/tut08_respt_cdf.m
% response-time-distribution-and-percentiles
model = Network('Model');

% Block 1: nodes
node{1} = Delay(model, 'Delay');
node{2} = Queue(model, 'Queue1', SchedStrategy.PS);

% Block 2: classes
jobclass{1} = ClosedClass(model, 'Class1', 5, node{1}, 0);
node{1}.setService(jobclass{1}, Exp(1.0));
node{2}.setService(jobclass{1}, Exp(0.5));

% Block 3: topology
model.link(Network.serialRouting(node{1},node{2}));
%%
% Block 4: solution
RDfluid = FLD(model,'seed',23000).getCdfRespT();
RDsim = JMT(model,'seed',23000,'samples',1e4).getCdfRespT();

% Plot results
if ~isempty(RDsim{2,1})
    semilogx(RDsim{2,1}(:,2),1-RDsim{2,1}(:,1),'r'); hold on;
    semilogx(RDfluid{2,1}(:,2),1-RDfluid{2,1}(:,1),'k--');
    legend('jmt-transient','fluid-steady','Location','Best');
    ylabel('Pr(T > t)'); xlabel('time t');
end
%%
% Compute CDF-derived scalar statistics
M = model.getNumberOfStations;
K = model.getNumberOfClasses;

AvgRespTfromCDFSim = zeros(M,K);
SqCoeffOfVariationRespTfromCDFSim = zeros(M,K);
for i=1:M
    for c=1:K
        if ~isempty(RDsim{i,c}) && size(RDsim{i,c},2) >= 2
            AvgRespTfromCDFSim(i,c) = diff(RDsim{i,c}(:,1))'*RDsim{i,c}(2:end,2);
            PowerMoment2_R(i,c) = diff(RDsim{i,c}(:,1))'*(RDsim{i,c}(2:end,2).^2);
            Variance_R(i,c) = PowerMoment2_R(i,c)-AvgRespTfromCDFSim(i,c)^2;
            SqCoeffOfVariationRespTfromCDFSim(i,c) = Variance_R(i,c)/AvgRespTfromCDFSim(i,c)^2;
        end
    end
end

AvgRespTfromCDFFluid = zeros(M,K);
SqCoeffOfVariationRespTfromCDFFluid = zeros(M,K);
for i=1:M
    for c=1:K
        if ~isempty(RDfluid{i,c}) && size(RDfluid{i,c},2) >= 2
            AvgRespTfromCDFFluid(i,c) = diff(RDfluid{i,c}(:,1))'*RDfluid{i,c}(2:end,2);
            PowerMoment2_R(i,c) = diff(RDfluid{i,c}(:,1))'*(RDfluid{i,c}(2:end,2).^2);
            Variance_R(i,c) = PowerMoment2_R(i,c)-AvgRespTfromCDFFluid(i,c)^2;
            SqCoeffOfVariationRespTfromCDFFluid(i,c) = Variance_R(i,c)/AvgRespTfromCDFFluid(i,c)^2;
        end
    end
end

AvgRespTfromCDFSim
AvgRespTfromCDFFluid
SqCoeffOfVariationRespTfromCDFSim
SqCoeffOfVariationRespTfromCDFFluid

File: python/examples/gettingstarted/tut08_respt_cdf.py
# %%
from line_solver import *
import numpy as np
GlobalConstants.set_verbose(VerboseLevel.STD)
import matplotlib.pyplot as plt
# %%
model = Network('Model')

# Block 1: nodes
node = np.empty(2, dtype=object)
node[0] = Delay(model, 'Delay')
node[1] = Queue(model, 'Queue1', SchedStrategy.PS)

# Block 2: classes
jobclass = np.empty(2, dtype=object)
jobclass[0] = ClosedClass(model, 'Class1', 5, node[0], 0)
node[0].set_service(jobclass[0], Exp(1.0))
node[1].set_service(jobclass[0], Exp(0.5))

# Block 3: topology
model.link(Network.serial_routing(node[0], node[1]))
# %%
# Block 4: solution
RDfluid = FLD(model).cdf_resp_t()
RDsim = JMT(model, seed=23000, samples=10000).cdf_resp_t()
# %%
# Plot results
import matplotlib.pyplot as plt

if RDsim[1][0] is not None and len(RDsim[1][0]) > 0:
    plt.figure(figsize=(10, 6))
    plt.semilogx(RDsim[1][0][:, 1], 1 - RDsim[1][0][:, 0], 'r-', label='jmt-transient')
    plt.semilogx(RDfluid[1][0][:, 1], 1 - RDfluid[1][0][:, 0], 'k--', label='fluid-steady')
    plt.legend(loc='best')
    plt.ylabel('Pr(T > t)')
    plt.xlabel('time t')
    plt.title('Response Time Distribution and Percentiles')
    plt.grid(True, alpha=0.3)
    plt.show(block=False)
    plt.close('all')
else:
    print("No simulation results available for plotting")
# %%
# Compute CDF-derived scalar statistics
M = model.getNumberOfStations()
K = model.getNumberOfClasses()

avg_respt_from_cdf_sim = []
sq_coeff_of_variation_resp_t_from_cdf_sim = []
for i in range(M):
    avg_respt_from_cdf_sim.append([])
    sq_coeff_of_variation_resp_t_from_cdf_sim.append([])
    for c in range(K):
        cdf_data = RDsim[i][c]
        if cdf_data is not None and len(cdf_data) > 1:
            mean_val = sum((cdf_data[j+1][0] - cdf_data[j][0]) * cdf_data[j+1][1]
                          for j in range(len(cdf_data)-1))
            power_moment_2 = sum((cdf_data[j+1][0] - cdf_data[j][0]) * (cdf_data[j+1][1] ** 2)
                                for j in range(len(cdf_data)-1))
            variance = power_moment_2 - mean_val ** 2
            scv = variance / (mean_val ** 2) if mean_val > 0 else 0
            avg_respt_from_cdf_sim[i].append(mean_val)
            sq_coeff_of_variation_resp_t_from_cdf_sim[i].append(scv)
        else:
            avg_respt_from_cdf_sim[i].append(0)
            sq_coeff_of_variation_resp_t_from_cdf_sim[i].append(0)

avg_respt_from_cdf_fluid = []
sq_coeff_of_variation_resp_t_from_cdf_fluid = []
for i in range(M):
    avg_respt_from_cdf_fluid.append([])
    sq_coeff_of_variation_resp_t_from_cdf_fluid.append([])
    for c in range(K):
        cdf_data = RDfluid[i][c]
        if cdf_data is not None and len(cdf_data) > 1:
            mean_val = sum((cdf_data[j+1][0] - cdf_data[j][0]) * cdf_data[j+1][1]
                          for j in range(len(cdf_data)-1))
            power_moment_2 = sum((cdf_data[j+1][0] - cdf_data[j][0]) * (cdf_data[j+1][1] ** 2)
                                for j in range(len(cdf_data)-1))
            variance = power_moment_2 - mean_val ** 2
            scv = variance / (mean_val ** 2) if mean_val > 0 else 0
            avg_respt_from_cdf_fluid[i].append(mean_val)
            sq_coeff_of_variation_resp_t_from_cdf_fluid[i].append(scv)
        else:
            avg_respt_from_cdf_fluid[i].append(0)
            sq_coeff_of_variation_resp_t_from_cdf_fluid[i].append(0)

print('\nAverage Response Time from CDF (Simulation):')
print(avg_respt_from_cdf_sim)
print('\nAverage Response Time from CDF (Fluid):')
print(avg_respt_from_cdf_fluid)
print('\nSquared Coefficient of Variation from CDF (Simulation):')
print(sq_coeff_of_variation_resp_t_from_cdf_sim)
print('\nSquared Coefficient of Variation from CDF (Fluid):')
print(sq_coeff_of_variation_resp_t_from_cdf_fluid)